src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -92,22 +96,27 @@ public void Start()
             // Start socket prior to applying the ConnectionAdapter
             _socket.ReadStart(_allocCallback, _readCallback, this);
 
-            if (_connectionAdapters.Count == 0)
-            {
-                _frame.Start();
-            }
-            else
+            // Dispatch to a thread pool so if the first read completes synchronously
+            // we won't be on IO thread
+            ThreadPool.Run(() =>
```

Comment:
Use ThreadPool.UnsafeRun. If it can throw, put a try/catch around this and log it.